preclustering of the remage outputs and refactor of output schemes / detectors (#331)
legend-exp/remage: docs/manual/output.md

Reviewer: @tdixon97 have a look

cannot contain forward slashes, so units lkike `m/s` are not possible to
represent directly. Instead, a backslash (`\`) is used to encode the division
symbol (for example: `velocity_in_m\s`).

## Germanium (HPGe) detectors

The Germanium output scheme handles the output from Germanium (HPGe) detectors, but would
also work for other solid state detectors (calorimeters).

Germanium detectors have sensitivity to the topology of event interactions via the pulse
shape and they also have a different response close to the detector electrodes. So
when simulating HPGe's it is advisable to save the information of the steps of
particles within the detector. Then "post-processing" software such as _reboost_
([docs](https://reboost.readthedocs.io/en/stable/)) can apply the detector response
model without repeating the computationally intensive simulation.

By default this output scheme writes out all steps in the registered sensitive
HPGe detectors. The following properties of each hit are recorded (by default):

> - _time_: The global time of the hit,

Reviewer: I prefer monospaced font here
```suggestion
- `time`: The global time of the hit,
```

- _particle_: the PDG code of the particle,
- _xloc_, _yloc_, _zloc_: the global position,
- _evtid_ : the index of the Geant4 event,
- _edep_: the deposited energy,
- _dist_to_surf_: the distance of the hit from the detector surface.

By default all floating point fields are saved with 64-bit (double) precision.
The precision of the energy and or position / distance fields can be reduced
to 32-bit with the macro commands:

> `/RMG/Output/Germanium/StoreSinglePrecisionPosition true`

Reviewer: Would be nice to directly link to the command reference every time (see #334)

Reviewer: Also, `true` is optional and not really making the command more clear here?

and,

`/RMG/Output/Germanium/StoreSinglePrecisionEnergy true`

> It is possible to also store the track id and parent track id of each step with:

Reviewer: what is the "track identifier"?

Author: track id

Reviewer: i meant maybe two words about where this id comes from

`/RMG/Output/Germanium/StoreTrackID true`

As mentioned earlier output schemes also provide a mechanism for filtering
events. One useful option is to only write out events in which energy

> was deposited in a Germanium detector. This is used since the

Reviewer: Is Germanium capital because it's the name of the sensitive detector type? Then maybe use monospaced or italic font?

> other detector systems (LAr, water Cherenkov etc.) often act a "vetos's",

Reviewer: ```suggestion
other detector systems (liquid argon, water Cherenkov etc.) often act a "vetos's",
```

so we are not interested in the energy deposited or steps if an event in the

> Germanium did not occur.

Reviewer: ```suggestion
germanium did not occur.
```

The macro commands:

> `/RMG/Output/Germanium/AddDetectorForEdepThreshold {DET_UID}`

Reviewer: this is the syntax for variable substitutions, I think it's confusing to use it here. just use DET_UID

photon tracks before simulating them if no energy was deposited in Germanium. This can be enabled
with the:

`/RMG/Output/Germanium/DiscardPhotonsIfNoGermaniumEdep`

macro command.

By default, the position saved for each step is the average of the pre and post-step point. This can
be controlled with the macro command:

`/RMG/Output/Germanium/StepPositionMode`

which can be set to `Average` (the default), `Both` (saves) also the pre and post steps, or `Pre`/`Post`.

:::{important}
For gammas the position saved is always that of the post-step, since all gamma interactions are discrete.
:::

Typically only steps where some energy was deposited are written out to disk, to control
this behaviour there is a macro command:

`/RMG/Output/Germanium/DiscardZeroEnergyHits`

Finally, it is possible to "pre-cluster" the steps, this is used to reduce the
amount of data written out to disk by combining steps very close together.
Since the surface region of a HPGe detector has different properties to the bulk
this clustering can be performed differently for surface and bulk hits
(see [data-reduction](#data-reduction-methods) for more details).

## Scintillator detectors

> This output scheme is used to record the steps in scintillation detectors (typically LAr),

Reviewer: ```suggestion
This output scheme is used to record the steps in scintillation detectors (typically liquid argon),
```

this is a calometric approach recording the energy deposited and steps. While the Optical
output scheme is instead used for recording the detected optical photons.
Most functionality is similar to the Germanium output scheme with a few exceptions.

- Unlike for Germanium detectors the distance to the detector surface is not calculated,
- The stacking possibility for optical tracks is not implemented,
- The velocity of the particles can be saved using the

> `/RMG/Output/Scintillator/StoreParticleVelocities`

Reviewer: here and in all the rest: put everything on the same line and link to the macro command reference

macro command.

## Data reduction methods

Often Geant4 takes steps much shorter than those that are meaningful in a HPGe or a
scintillation detector. For example the typical dimension of charge clouds

> produced by interactions in Germanium are 1-2 mm, so we are not sensitive to tracking at um levels.

Reviewer: ```suggestion
produced by interactions in Germanium are 1-2 mm, so we are not sensitive to tracking at the micrometer level.
```

To reduce the file size while retaining the useful information for computing observables of interest

> we have implemented some "pre-clustering" routines. These routines combine together steps that are very
> very close together.

Reviewer: ```suggestion
we have implemented some "pre-clustering" routines. These routines combine together steps that are very
very very very very very very very very very close together.
```

:::{note}
The aim of this (pre)-clustering is only to make a minimal reduction of information which cannot be useful!
Further, more aggressive clustering may be needed for some applications.
:::

In order to have an efficient algorithm for pre-clustering we take use a "within-track" approach,
this clusters only steps in the same `G4Track`, with some exceptions for very low energy tracks.
In this way we only have to iterate through the steps in each event once. This also means the rows in our output are still interpretable
with steps in the detector (just with a larger step length). The clustering is handled by the
function:

```C++

>   RMGDetectorHitsCollection* pre_cluster_hits(const RMGDetectorHitsCollection* hits,

Reviewer: also here link to the c++ api, see #334

      ClusterPars cluster_pars, bool has_distance_to_surface, bool has_velocity);
```

> This takes in the pointer to the original `RMGDetectorHitsCollection` returning a pointer

Reviewer: also here missing API reference

to a new collection of clustered hits.

:::{note}
This design makes it easy to include additional clustering algorithms, a similar function just needs to be written.
:::
Pre-clustering can be enabled with the macro command:

`/RMG/Output/Germanium/PreClusterOutputs True`

and similar for the `Scintillator` output scheme.

> _remage_ first organises the hits by track id. Some processes in _Geant4_ produce

Reviewer: ```suggestion
_remage_ first organises the hits by track id. Some processes in Geant4 produce
```
